src/v/security/scram_algorithm.cc

```diff
@@ -109,6 +123,33 @@ ctre_parse_client_first(std::string_view message) {
     };
 }
 
+static inline std::optional<server_first_match>
+ctre_parse_server_first(std::string_view message) {
+    auto match = ctre::match<server_first_message_re>(message);
+    if (unlikely(!match)) {
+        return std::nullopt;
+    }
+
+    int iterations; // NOLINT
+    auto i_str = match.get<4>().to_view();
+    auto res = std::from_chars(
+      i_str.data(), i_str.data() + i_str.size(), iterations);
+
+    // very unlikely since the regex should reject before this
+    if (unlikely(res.ec != std::errc())) {
+        throw std::runtime_error(fmt_with_ctx(
```

Comment:
this method is confusing to me. it 

1. uses optional
2. uses exceptions

Shouldn't we just use optional and log error ?

Reply:
yeh they should all throw instead of returning a nullopt. they are only used in constructors so there isn't any way to bubble up any information unless we refactor to use factories of some sort

Reply:
shouldn't it be composable then 

```
static optional<foo> throw_if_nullopt(std::optional<foo> f) { if !f throw; return f;}

```

Reply:
anyway,that's a nit.

Reply:
nullopt communicates that the message didn't match the regex while exceptions handle all the other cases like integer/string conversion etc... anyway this is an internal helper api local to the file.